File: backend/app/providers/coingecko.py
"""Cliente compartilhado do CoinGecko `/simple/price`.

Bitcoin e currencies dividem o mesmo bucket IP. A API *keyless* é ~10–30
chamadas/min (dinâmico) e a própria CoinGecko cacheia 1–5 min — pollar a
cada 60 s só gasta o limite. Ver https://docs.coingecko.com/docs/keyless-public-api
"""

from __future__ import annotations

import threading
import time
import urllib.parse
from collections.abc import Sequence
from typing import Any

from app.http_util import HttpError, http_json, is_rate_limit

SIMPLE_PRICE_URL = "https://api.coingecko.com/api/v3/simple/price"

# A API pública já entrega dado de 1–5 min. 5 min = ~12 req/h, bem abaixo do teto.
TTL_S = 300
_MIN_BACKOFF_S = 120.0

_lock = threading.Condition(threading.Lock())
_inflight = False
# querystring -> (monotonic, payload)
_cache: dict[str, tuple[float, dict[str, Any]]] = {}
_backoff_until = 0.0


def reset() -> None:
    global _inflight, _backoff_until
    with _lock:
        _cache.clear()
        _inflight = False
        _backoff_until = 0.0


def _query(ids: Sequence[str], vs: Sequence[str]) -> str | None:
    clean_ids = sorted({str(i).strip().lower() for i in ids if str(i).strip()})
    clean_vs = sorted({str(v).strip().lower() for v in vs if str(v).strip()})
    if not clean_ids or not clean_vs:
        return None
    return urllib.parse.urlencode({"ids": ",".join(clean_ids), "vs_currencies": ",".join(clean_vs)})


def fetch_simple_price(ids: Sequence[str], vs_currencies: Sequence[str]) -> dict[str, Any]:
    """Devolve o JSON do CoinGecko, com TTL, coalescência e last-good em 429."""
    qs = _query(ids, vs_currencies)
    if not qs:
        return {}

    global _inflight, _backoff_until
    stale: dict[str, Any] | None = None
    with _lock:
        while True:
            now = time.monotonic()
            entry = _cache.get(qs)
            if entry is not None and (now - entry[0]) < TTL_S:
                return entry[1]
            if now < _backoff_until and entry is not None:
                return entry[1]
            if _inflight:
                _lock.wait(timeout=20.0)
                continue
            _inflight = True
            stale = entry[1] if entry is not None else None
            break

    try:
        data = http_json(f"{SIMPLE_PRICE_URL}?{qs}", timeout=15.0)
        if not isinstance(data, dict):
            raise RuntimeError("resposta inesperada da cotação")
        with _lock:
            _cache[qs] = (time.monotonic(), data)
            _backoff_until = 0.0
        return data
    except Exception as exc:  # noqa: BLE001
        if is_rate_limit(exc) and stale is not None:
            wait = _MIN_BACKOFF_S
            if isinstance(exc, HttpError) and exc.retry_after_s:
                wait = max(wait, exc.retry_after_s)
            with _lock:
                _backoff_until = time.monotonic() + wait
            return stale
        raise
    finally:
        with _lock:
            _inflight = False
            _lock.notify_all()

File: backend/src/providers/coingecko.ts
/**
 * Cliente compartilhado do CoinGecko /simple/price.
 * Port of backend-python-legacy/app/providers/coingecko.py
 */
import { httpJson, HttpError, isRateLimit } from "../httpClient.js";

export const SIMPLE_PRICE_URL = "https://api.coingecko.com/api/v3/simple/price";
export const TTL_S = 300;
const MIN_BACKOFF_S = 120.0;

type CacheEntry = { at: number; payload: Record<string, unknown> };

const cache = new Map<string, CacheEntry>();
let backoffUntil = 0; // epoch ms via Date.now()
let inflightPromise: Promise<Record<string, unknown>> | null = null;
let inflightKey: string | null = null;

export function reset(): void {
  cache.clear();
  inflightPromise = null;
  inflightKey = null;
  backoffUntil = 0;
}

function query(ids: string[], vs: string[]): string | null {
  const cleanIds = [...new Set(ids.map((i) => String(i).trim().toLowerCase()).filter(Boolean))].sort();
  const cleanVs = [...new Set(vs.map((v) => String(v).trim().toLowerCase()).filter(Boolean))].sort();
  if (cleanIds.length === 0 || cleanVs.length === 0) return null;
  return new URLSearchParams({ ids: cleanIds.join(","), vs_currencies: cleanVs.join(",") }).toString();
}

export async function fetchSimplePrice(
  ids: string[],
  vsCurrencies: string[],
  provider = "CRYPTO",
): Promise<Record<string, unknown>> {
  const qs = query(ids, vsCurrencies);
  if (!qs) return {};

  const now = Date.now();
  const entry = cache.get(qs);
  if (entry && (now - entry.at) / 1000 < TTL_S) return entry.payload;
  if (now < backoffUntil && entry) return entry.payload;

  // coalescence: if already inflight for same qs, await it
  if (inflightPromise && inflightKey === qs) {
    try {
      const data = await inflightPromise;
      return data;
    } catch {
      // fall through to retry (if last-good available will be returned below via 429 handling)
      // but for coalescence we need to attempt new fetch if inflight failed without stale
      // For simplicity, continue to start new fetch after catch
    }
  }
  // if any inflight for different key, we still serialize per key? Python uses global _inflight lock.
  // To mimic, if any inflight exists, wait for it then re-check cache.
  if (inflightPromise) {
    try {
      await inflightPromise;
    } catch {}
    const after = cache.get(qs);
    const now2 = Date.now();
    if (after && (now2 - after.at) / 1000 < TTL_S) return after.payload;
    if (now2 < backoffUntil && after) return after.payload;
  }

  const stale = entry ? entry.payload : null;

  const url = `${SIMPLE_PRICE_URL}?${qs}`;

  const promise = (async () => {
    const data = await httpJson(url, { timeout: 15.0, provider });
    if (data === null || typeof data !== "object" || Array.isArray(data)) throw new Error("resposta inesperada da cotação");
    cache.set(qs, { at: Date.now(), payload: data as Record<string, unknown> });
    backoffUntil = 0;
    return data as Record<string, unknown>;
  })();

  inflightPromise = promise as Promise<Record<string, unknown>>;
  inflightKey = qs;

  try {
    const result = await promise;
    return result;
  } catch (exc) {
    if (isRateLimit(exc) && stale !== null) {
      let wait = MIN_BACKOFF_S;
      if (exc instanceof HttpError && exc.retryAfterS !== null) {
        wait = Math.max(wait, exc.retryAfterS);
      }
      backoffUntil = Date.now() + wait * 1000;
      return stale;
    }
    throw exc;
  } finally {
    if (inflightPromise === promise) {
      inflightPromise = null;
      inflightKey = null;
    }
  }
}

export const fetch_simple_price = fetchSimplePrice;
